document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/transformer/PostgresSelectionQueryTransformer.java

```diff
@@ -29,42 +33,103 @@
  *
  * This is the similar behavior supported in our other document store implementation (e.g Mongo)
  * */
-public class PostgresSelectionQueryTransformer
-    implements QueryTransformer, SelectTypeExpressionVisitor {
+public class PostgresSelectionQueryTransformer implements QueryTransformer {
 
   @Override
   public Query transform(Query query) {
-    // no-op if group by clause exits
-    if (!query.getAggregations().isEmpty()) return query;
+    LocalSelectTypeAggregationExpressionSelector aggregationExpressionSelector =
+        new LocalSelectTypeAggregationExpressionSelector();
+    Boolean noAggregation =
+        query.getSelections().stream()
+            .filter(
+                selectionSpec ->
+                    selectionSpec.getExpression().accept(aggregationExpressionSelector))
+            .findAny()
+            .isEmpty();
+    if (noAggregation) return query;
+
+    // get all identifier of group by clause
+    LocalGroupByIdentifierExpressionSelector groupByIdentifierExpressionSelector =
+        new LocalGroupByIdentifierExpressionSelector();
+    Set<GroupTypeExpression> groupByIdentifierExpressions =
+        query.getAggregations().stream()
+            .filter(exp -> exp.accept(groupByIdentifierExpressionSelector))
+            .collect(Collectors.toUnmodifiableSet());
 
-    // check for all selections, remove non-aggregated selections.
+    // keep only matching identifier expression with group by along with rest.
+    LocalSelectTypeIdentifierExpressionSelector identifierExpressionSelector =
+        new LocalSelectTypeIdentifierExpressionSelector();
     List<SelectionSpec> finalSelectionSpecs =
         query.getSelections().stream()
-            .filter(selectionSpec -> selectionSpec.getExpression().accept(this))
+            .filter(
+                selectionSpec -> {
+                  SelectTypeExpression expression = selectionSpec.getExpression();
+                  return !((boolean) expression.accept(identifierExpressionSelector))
+                      || groupByIdentifierExpressions.contains(expression);
+                })
             .collect(Collectors.toUnmodifiableList());
 
-    return finalSelectionSpecs.size() > 0
+    return finalSelectionSpecs.size() != query.getSelections().size()
         ? new TransformedQueryBuilder(query).setSelections(finalSelectionSpecs).build()
         : query;
   }
 
-  @Override
-  public Boolean visit(AggregateExpression expression) {
-    return true;
-  }
+  private static class LocalSelectTypeAggregationExpressionSelector
```

Comment:
If this is just returning the instance type, do need a separate visitor hierarchy? Could we just do using instance of?

Reply:
I think, this is a more readable pattern.